update-engine/src/buffer.rs

```diff
@@ -253,153 +277,409 @@ impl<S: StepSpec> EventStore<S> {
+                    self.event_tree.add_edge(
+                        EventTreeNode::Step(parent_key),
+                        EventTreeNode::Root(nested_new_execution.execution_id),
+                        (),
+                    );
+                }
+
+                new_execution = actions.new_execution;
+                (actions.step_key, actions.progress_key)
             }
-            StepEventKind::NoStepsDefined | StepEventKind::Unknown => None,
-        }
+            StepEventKind::NoStepsDefined | StepEventKind::Unknown => {
+                (None, None)
+            }
+        };
+
+        RecurseActions { new_execution, step_key, progress_key }
     }
 
-    fn recurse_for_progress_event<S2: StepSpec>(
-        &mut self,
+    fn step_key_for_progress_event<S2: StepSpec>(
         event: &ProgressEvent<S2>,
-        parent_node: Option<StepKey>,
     ) -> Option<StepKey> {
         match &event.kind {
             ProgressEventKind::WaitingForProgress { step, .. }
             | ProgressEventKind::Progress { step, .. } => {
-                let key = self.add_node(event.execution_id, step.info.index)?;
-                if let Some(parent_node) = parent_node {
-                    self.event_tree.add_edge(parent_node, key, ());
-                }
+                let key = StepKey {
+                    execution_id: event.execution_id,
+                    index: step.info.index,
+                };
                 Some(key)
             }
-            ProgressEventKind::Nested { step, event: nested_event, .. } => {
-                // Add a node for this event.
-                let parent =
-                    self.add_node(event.execution_id, step.info.index)?;
-
-                // Add nodes for nested events.
-                self.recurse_for_progress_event(nested_event, Some(parent))
+            ProgressEventKind::Nested { event: nested_event, .. } => {
+                Self::step_key_for_progress_event(nested_event)
             }
             ProgressEventKind::Unknown => None,
         }
     }
 
-    fn add_node(
-        &mut self,
-        execution_id: ExecutionId,
-        index: usize,
-    ) -> Option<StepKey> {
-        Some(self.event_tree.add_node(StepKey { execution_id, index }))
+    fn add_root_node(&mut self, execution_id: ExecutionId) -> EventTreeNode {
+        self.event_tree.add_node(EventTreeNode::Root(execution_id))
     }
 
-    fn mark_event_key_completed(&mut self, key: StepKey) {
-        // Remove this node and anything reachable from it.
-        let mut dfs = DfsPostOrder::new(&self.event_tree, key);
+    fn add_step_node(&mut self, key: StepKey) -> EventTreeNode {
+        self.event_tree.add_node(EventTreeNode::Step(key))
+    }
+
+    fn mark_step_key_completed(
+        &mut self,
+        root_key: StepKey,
+        info: CompletionInfo,
+    ) {
+        if let Some(value) = self.map.get_mut(&root_key) {
+            // Completion status only applies to the root key.
```

Comment:
I found this comment misleading at first - my misinterpretation was "why does completion only apply to the root key if we're marking anything reachable from it complete too?", missing the importance of `status`. Maybe expand with something like this?

```rust
// Completion status only applies to the root key. Nodes
// reachable from this node are still marked as complete but
// without status.
```

Reply:
Good catch, thanks.